### PR TITLE
Kuznetsov home work#9

### PR DESCRIPTION
**Pull request by Muao**

### CODE REVIEW

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/controllers/LogOutController.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+@Controller
+@RequestMapping(value = "/logout")
+public class LogOutController {
+
+
+    @RequestMapping(method = POST, value = "")
+    public void logOut(ServletRequest servletRequest, ServletResponse servletResponse) throws IOException {
+
+        HttpServletRequest req = (HttpServletRequest) servletRequest;
+        HttpServletResponse resp = (HttpServletResponse) servletResponse;
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/controllers/LogOutController.java]: You are using controller as plain servlet. You can receive Session as parameter and return String with "redirect:/"

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/controllers/LoginController.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    @Autowired
+    private QuizDaoHibernate quizDao;
+
+    @Autowired
+    private UsersEntity usersEntity;
+
+    @RequestMapping(method = GET, value = "")
+    public void getLoginView(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+
+        RequestDispatcher loginDispatcher = req.getRequestDispatcher("/view/login.jsp");
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/controllers/LoginController.java]: The same here

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/dao/impl/QuizDaoHibernate.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        Query query = session.createQuery("from " + QuizzesEntity.class.getName());
+        return query.list();
+    }
+
+    public SubjectsEntity getSubjectsFromDb(String subject) {
+
+        Session session = HibernateUtil.getSessionFactory().openSession();
+        Transaction transaction = session.beginTransaction();
+        SubjectsEntity subjectsEntity = session.byNaturalId(SubjectsEntity.class).using("subject", subject).load();
+        transaction.commit();
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/dao/impl/QuizDaoHibernate.java]: Get can work without transaction

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/dao/impl/QuizDaoHibernate.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import org.hibernate.query.Query;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class QuizDaoHibernate implements QuizDao {
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/dao/impl/QuizDaoHibernate.java]: It's better to separate dao methods per entity.

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/entities/QuestionsEntity.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import javax.persistence.*;
+import java.util.Objects;
+
+@Entity
+@Table(name = "questions", schema = "quiz")
+public class QuestionsEntity {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int id;
+    @NaturalId
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/entities/QuestionsEntity.java]: Bad idea twice
1. ThemeId is not natural id
2. In Hibernate Question must contain Theme, not only id

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/entities/QuizzesEntity.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
+package com.kuznetsov.entities;
+
+import org.springframework.stereotype.Component;
+
+import javax.persistence.*;
+import java.util.Objects;
+
+@Entity
+@Table(name = "quizzes", schema = "quiz")
+@Component
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/entities/QuizzesEntity.java]: Why you use component at entity?

**Review by dinamio — CHANGES_REQUESTED** [review]
Please, change approach to controllers, and hibernate mapping

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/entities/QuizzesEntity.java (hunk 6 above)
Comment (Muao) [reply kuznetsov/quizzes/src/main/java/com/kuznetsov/entities/QuizzesEntity.java]: It left from the previous implementation

File: kuznetsov/quizzes/src/main/java/com/kuznetsov/controllers/QuizController.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    public String addNewQuiz(@ModelAttribute("dataFromForm") QuizDataFromForm dataFromForm, HttpServletRequest req) {
+        String theme = dataFromForm.getTheme();
+
+        if (theme != null) {
+            String subject = dataFromForm.getSubject();
+            String sessionLogin = (String) req.getSession().getAttribute("login");
+
+            Map<String, Byte> questionMap = new HashMap<>(questionAggregator.createQuestionsMap(req));
+            services.addNewQuiz(subject, theme, sessionLogin, questionMap);
+            req.setAttribute("list", services.getAllQuizzes());
```
Comment (dinamio) [inline kuznetsov/quizzes/src/main/java/com/kuznetsov/controllers/QuizController.java:52]: This attribute idealogically must be in Model. But for now it's okay

**Review by dinamio — APPROVED** [review]
Looks good

**Pull request merged**